Missing cluster finilazers due to possible data race

**Steps to reproduce:**
 * I create openstack cluster
 * I see cloud [openstack InitializeCloudProvider function](https://github.com/kubermatic/kubermatic/blob/master/pkg/provider/cloud/openstack/provider.go#L158) created resources in openstack
 * When later I check cluster finalizers by running kubectl describe
 * I see [the finalizers](https://github.com/kubermatic/kubermatic/blob/master/pkg/provider/cloud/openstack/provider.go#{L45:L58}) are missing
 
I debugged and found that finalizers were added and deleted shortly after. I found that finalizers added by [openstack InitializeCloudProvider function](https://github.com/kubermatic/kubermatic/blob/master/pkg/provider/cloud/openstack/provider.go#L158) were deleted by following update here https://github.com/kubermatic/kubermatic/blob/master/pkg/controller/seed-controller-manager/kubernetes/pending.go#L90

I was able to fix the bug by reverting reverting changes introduced #4578

**Kubermatic version:** `v2.16.5`

I wrote a program to reproduce data loss. It performs two patches to add finalizers `foo` and `bar`. If you execute it the final result contains `bar` finalizer from last patch, but doesn't contain finalizer `foo`.
```go
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"os"

	"github.com/pkg/errors"
	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
	ctrlruntime "sigs.k8s.io/controller-runtime"
	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/manager"
)

func main() {
	clusterName := flag.String("cluster", "", "cluster name")
	flag.Parse()
	if *clusterName == "" {
		flag.Usage()
		os.Exit(1)
	}

	cfg := ctrlruntime.GetConfigOrDie()
	mgr, err := manager.New(cfg, manager.Options{})
	if err != nil {
		panic(err)
	}
	go func() {
		if err := mgr.GetCache().Start(context.Background()); err != nil {
			panic(err)
		}
	}()

	client := mgr.GetClient()

	if ok := mgr.GetCache().WaitForCacheSync(ctrlruntime.SetupSignalHandler()); !ok {
		panic("cache is not ready")
	}

	var cluster kubermaticv1.Cluster
	if err := client.Get(context.Background(), ctrlruntimeclient.ObjectKey{"kubermatic", *clusterName}, &cluster); err != nil {
		panic(err)
	}
	cluster.SetManagedFields(nil)

	var clusterCopy kubermaticv1.Cluster
	if err := client.Get(context.Background(), ctrlruntimeclient.ObjectKey{"kubermatic", *clusterName}, &clusterCopy); err != nil {
		panic(err)
	}
	clusterCopy.SetManagedFields(nil)

	if err := patch(client, &cluster, "foo"); err != nil {
		log.Panicf("foo: %v", err)
	}
	if err := patch(client, &clusterCopy, "bar"); err != nil {
		log.Panicf("bar: %v", err)
	}
	fmt.Println("done")
}

func patch(client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster, finalizers ...string) error {
	oldCluster := cluster.DeepCopy()
	kuberneteshelper.AddFinalizer(cluster, finalizers...)
	if err := client.Patch(context.Background(), cluster, ctrlruntimeclient.MergeFrom(oldCluster)); err != nil {
		return errors.Wrap(err, "patch cluster")
	}
	return nil
}
```

---

Here is how server-side apply would look. Here I patch finalizers by using `PartialObjectMetadata` because if I use `kubermaticv1.Cluster` I locks for more fields than I need.
```go
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"os"

	"github.com/pkg/errors"
	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	ctrlruntime "sigs.k8s.io/controller-runtime"
	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
	"sigs.k8s.io/controller-runtime/pkg/manager"
)

func main() {
	clusterName := flag.String("cluster", "", "cluster name")
	flag.Parse()
	if *clusterName == "" {
		flag.Usage()
		os.Exit(1)
	}

	cfg := ctrlruntime.GetConfigOrDie()
	mgr, err := manager.New(cfg, manager.Options{})
	if err != nil {
		panic(err)
	}
	go func() {
		if err := mgr.GetCache().Start(context.Background()); err != nil {
			panic(err)
		}
	}()

	client := mgr.GetClient()

	if ok := mgr.GetCache().WaitForCacheSync(ctrlruntime.SetupSignalHandler()); !ok {
		panic("cache is not ready")
	}

	var cluster kubermaticv1.Cluster
	if err := client.Get(context.Background(), ctrlruntimeclient.ObjectKey{"kubermatic", *clusterName}, &cluster); err != nil {
		panic(err)
	}
	cluster.SetManagedFields(nil)

	var clusterCopy kubermaticv1.Cluster
	if err := client.Get(context.Background(), ctrlruntimeclient.ObjectKey{"kubermatic", *clusterName}, &clusterCopy); err != nil {
		panic(err)
	}
	clusterCopy.SetManagedFields(nil)

	if err := patch(client, &cluster, "foo-controller", "foo"); err != nil {
		log.Panicf("foo: %v", err)
	}
	if err := patch(client, &clusterCopy, "bar-controller", "bar"); err != nil {
		log.Panicf("bar: %v", err)
	}
	fmt.Println("done")
}

func patch(client ctrlruntimeclient.Client, cluster *kubermaticv1.Cluster, controller string, finalizers ...string) error {
	patch := &metav1.PartialObjectMetadata{
		TypeMeta: cluster.TypeMeta,
		ObjectMeta: metav1.ObjectMeta{
			Name:       cluster.Name,
			Namespace:  cluster.Namespace,
			Finalizers: cluster.Finalizers,
		},
	}
	owner := ctrlruntimeclient.FieldOwner(controller)
	patch.SetManagedFields(nil)
	// Claim shared ownership
	if err := client.Patch(context.Background(), patch, ctrlruntimeclient.Apply, owner); err != nil {
		return errors.Wrap(err, "claim shared ownership")
	}
	// Patch finalizers
	kuberneteshelper.AddFinalizer(patch, finalizers...)
	patch.SetManagedFields(nil)
	if err := client.Patch(context.Background(), patch, ctrlruntimeclient.Apply, owner, ctrlruntimeclient.ForceOwnership); err != nil {
		return errors.Wrap(err, "add finalizers")
	}
	return nil
}
```

This results with following error, which I think is what we need.
```
panic: bar: claim shared ownership: Apply failed with 1 conflict: conflict with "foo-controller": .metadata.finalizers
```

WDYT @irozzo-1A ?

--- a/pkg/controller/master-controller-manager/usersshkeyssynchronizer/controller.go
+++ b/pkg/controller/master-controller-manager/usersshkeyssynchronizer/controller.go
@@ -208,8 +208,8 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, requ
 	oldCluster := cluster.DeepCopy()
 	if !kubernetes.HasFinalizer(cluster, UserSSHKeysClusterIDsCleanupFinalizer) {
 		kubernetes.AddFinalizer(cluster, UserSSHKeysClusterIDsCleanupFinalizer)
-		if err := seedClient.Patch(ctx, cluster, ctrlruntimeclient.MergeFrom(oldCluster)); err != nil {
-			return fmt.Errorf("failed adding %s finalizer: %v", UserSSHKeysClusterIDsCleanupFinalizer, err)
+		if err := seedClient.Patch(ctx, cluster, ctrlruntimeclient.MergeFromWithOptions(oldCluster, ctrlruntimeclient.MergeFromWithOptimisticLock{})); err != nil {
+			return fmt.Errorf("failed adding %s finalizer: %w", UserSSHKeysClusterIDsCleanupFinalizer, err)
 		}
 	}
 

--- a/pkg/controller/seed-controller-manager/backup/backup_controller.go
+++ b/pkg/controller/seed-controller-manager/backup/backup_controller.go
@@ -289,8 +289,8 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, clus
 	if !kuberneteshelper.HasFinalizer(cluster, cleanupFinalizer) && !r.disabled {
 		oldCluster := cluster.DeepCopy()
 		kuberneteshelper.AddFinalizer(cluster, cleanupFinalizer)
-		if err := r.Patch(ctx, cluster, ctrlruntimeclient.MergeFrom(oldCluster)); err != nil {
-			return fmt.Errorf("failed to update cluster after adding cleanup finalizer: %v", err)
+		if err := r.Patch(ctx, cluster, ctrlruntimeclient.MergeFromWithOptions(oldCluster, ctrlruntimeclient.MergeFromWithOptimisticLock{})); err != nil {
+			return fmt.Errorf("failed to update cluster after adding cleanup finalizer: %w", err)
 		}
 	}
 

--- a/pkg/controller/seed-controller-manager/cloud/cloud_controller.go
+++ b/pkg/controller/seed-controller-manager/cloud/cloud_controller.go
@@ -188,6 +188,12 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, clus
 	}
 
 	if _, err := prov.InitializeCloudProvider(cluster, r.updateCluster); err != nil {
+		if kerrors.IsConflict(err) {
+			// In case of conflict we just re-enqueue the item for later
+			// processing without returning an error.
+			r.log.Infow("failed to add finalizer to cluster", "error", err)
+			return &reconcile.Result{Requeue: true}, nil
+		}
 		return nil, fmt.Errorf("failed cloud provider init: %v", err)
 	}
 
@@ -248,7 +254,7 @@ func (r *Reconciler) migrateAWSMultiAZ(ctx context.Context, cluster *kubermaticv
 	return nil
 }
 
-func (r *Reconciler) updateCluster(name string, modify func(*kubermaticv1.Cluster)) (*kubermaticv1.Cluster, error) {
+func (r *Reconciler) updateCluster(name string, modify func(*kubermaticv1.Cluster), options ...provider.UpdaterOption) (*kubermaticv1.Cluster, error) {
 	cluster := &kubermaticv1.Cluster{}
 	if err := r.Get(context.Background(), types.NamespacedName{Name: name}, cluster); err != nil {
 		return nil, err
@@ -258,7 +264,14 @@ func (r *Reconciler) updateCluster(name string, modify func(*kubermaticv1.Cluste
 	if reflect.DeepEqual(oldCluster, cluster) {
 		return cluster, nil
 	}
-	return cluster, r.Patch(context.Background(), cluster, ctrlruntimeclient.MergeFrom(oldCluster))
+	opts := (&provider.UpdaterOptions{}).Apply(options...)
+	var patch ctrlruntimeclient.Patch
+	if opts.OptimisticLock {
+		patch = ctrlruntimeclient.MergeFromWithOptions(oldCluster, ctrlruntimeclient.MergeFromWithOptimisticLock{})
+	} else {
+		patch = ctrlruntimeclient.MergeFrom(oldCluster)
+	}
+	return cluster, r.Patch(context.Background(), cluster, patch)
 }
 
 func (r *Reconciler) getGlobalSecretKeySelectorValue(configVar *providerconfig.GlobalSecretKeySelector, key string) (string, error) {

--- a/pkg/controller/seed-controller-manager/kubernetes/cluster_controller.go
+++ b/pkg/controller/seed-controller-manager/kubernetes/cluster_controller.go
@@ -29,6 +29,7 @@ import (
 	controllerutil "k8c.io/kubermatic/v2/pkg/controller/util"
 	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
 	kubermaticv1helper "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1/helper"
+	kuberneteshelper "k8c.io/kubermatic/v2/pkg/kubernetes"
 	"k8c.io/kubermatic/v2/pkg/provider"
 	"k8c.io/kubermatic/v2/pkg/resources/certificates"
 	"k8c.io/kubermatic/v2/pkg/version/kubermatic"
@@ -38,7 +39,7 @@ import (
 	corev1 "k8s.io/api/core/v1"
 	policyv1beta1 "k8s.io/api/policy/v1beta1"
 	rbacv1 "k8s.io/api/rbac/v1"
-	kubeapierrors "k8s.io/apimachinery/pkg/api/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	autoscalingv1beta2 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
@@ -217,7 +218,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 
 	cluster := &kubermaticv1.Cluster{}
 	if err := r.Get(ctx, request.NamespacedName, cluster); err != nil {
-		if kubeapierrors.IsNotFound(err) {
+		if kerrors.IsNotFound(err) {
 			log.Debug("Could not find cluster")
 			return reconcile.Result{}, nil
 		}
@@ -260,7 +261,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
 	// synchronize cluster.status.health for Kubernetes clusters
 	if err := r.syncHealth(ctx, cluster); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to sync health: %w", err)
 	}
 
 	if cluster.DeletionTimestamp != nil {
@@ -284,7 +285,7 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, clus
 		if updateErr != nil {
 			return nil, fmt.Errorf("failed to set the cluster error: %v", updateErr)
 		}
-		return nil, err
+		return nil, fmt.Errorf("failed to reconcile cluster: %w", err)
 	}
 
 	if err := r.clearClusterError(ctx, cluster); err != nil {
@@ -294,13 +295,29 @@ func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, clus
 	return res, nil
 }
 
-func (r *Reconciler) updateCluster(ctx context.Context, cluster *kubermaticv1.Cluster, modify func(*kubermaticv1.Cluster)) error {
+func (r *Reconciler) updateCluster(ctx context.Context, cluster *kubermaticv1.Cluster, modify func(*kubermaticv1.Cluster), opts ...ctrlruntimeclient.MergeFromOption) error {
 	oldCluster := cluster.DeepCopy()
 	modify(cluster)
 	if reflect.DeepEqual(oldCluster, cluster) {
 		return nil
 	}
-	return r.Patch(ctx, cluster, ctrlruntimeclient.MergeFrom(oldCluster))
+
+	return r.Patch(ctx, cluster, ctrlruntimeclient.MergeFromWithOptions(oldCluster, opts...))
+}
+
+func (r *Reconciler) AddFinalizers(ctx context.Context, cluster *kubermaticv1.Cluster, finalizers ...string) (*reconcile.Result, error) {
+	if err := r.updateCluster(ctx, cluster, func(c *kubermaticv1.Cluster) {
+		kuberneteshelper.AddFinalizer(c, finalizers...)
+	}, ctrlruntimeclient.MergeFromWithOptimisticLock{}); err != nil {
+		if !kerrors.IsConflict(err) {
+			return nil, fmt.Errorf("failed to add finalizers %v: %w", finalizers, err)
+		}
+		// In case of conflict we just re-enqueue the item for later
+		// processing without returning an error.
+		r.log.Infow("failed to add finalizers", "error", err, "finalizers", finalizers)
+		return &reconcile.Result{Requeue: true}, nil
+	}
+	return &reconcile.Result{}, nil
 }
 
 func (r *Reconciler) updateClusterError(ctx context.Context, cluster *kubermaticv1.Cluster, reason kubermaticv1.ClusterStatusError, message string) error {

--- a/pkg/controller/seed-controller-manager/kubernetes/pending.go
+++ b/pkg/controller/seed-controller-manager/kubernetes/pending.go
@@ -44,10 +44,12 @@ func (r *Reconciler) reconcileCluster(ctx context.Context, cluster *kubermaticv1
 	}
 
 	if !kuberneteshelper.HasFinalizer(cluster, kubermaticapiv1.EtcdBackupConfigCleanupFinalizer) {
-		if err := r.updateCluster(ctx, cluster, func(c *kubermaticv1.Cluster) {
-			kuberneteshelper.AddFinalizer(c, kubermaticapiv1.EtcdBackupConfigCleanupFinalizer)
-		}); err != nil {
+		res, err := r.AddFinalizers(ctx, cluster, kubermaticapiv1.EtcdBackupConfigCleanupFinalizer)
+		if err != nil {
 			return nil, err
+		}
+		if !res.IsZero() {
+			return res, nil
 		}
 	}
 
@@ -61,6 +63,7 @@ func (r *Reconciler) reconcileCluster(ctx context.Context, cluster *kubermaticv1
 		return nil, err
 	}
 
+	var finalizers []string
 	if cluster.Status.ExtendedHealth.Apiserver == kubermaticv1.HealthStatusUp {
 		// Controlling of user-cluster resources
 		reachable, err := r.clusterIsReachable(ctx, cluster)
@@ -75,32 +78,20 @@ func (r *Reconciler) reconcileCluster(ctx context.Context, cluster *kubermaticv1
 		// Only add the node deletion finalizer when the cluster is actually running
 		// Otherwise we fail to delete the nodes and are stuck in a loop
 		if !kuberneteshelper.HasFinalizer(cluster, kubermaticapiv1.NodeDeletionFinalizer) {
-			err = r.updateCluster(ctx, cluster, func(c *kubermaticv1.Cluster) {
-				kuberneteshelper.AddFinalizer(c, kubermaticapiv1.NodeDeletionFinalizer)
-			})
-			if err != nil {
-				return nil, err
-			}
+			finalizers = append(finalizers, kubermaticapiv1.NodeDeletionFinalizer)
 		}
 
 	}
 
 	if !kuberneteshelper.HasFinalizer(cluster, kubermaticapiv1.KubermaticConstraintCleanupFinalizer) {
-		err := r.updateCluster(ctx, cluster, func(c *kubermaticv1.Cluster) {
-			kuberneteshelper.AddFinalizer(c, kubermaticapiv1.KubermaticConstraintCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
-		}
+		finalizers = append(finalizers, kubermaticapiv1.KubermaticConstraintCleanupFinalizer)
+	}
+	if !kuberneteshelper.HasFinalizer(cluster, kubermaticapiv1.ClusterRoleBindingsCleanupFinalizer) {
+		finalizers = append(finalizers, kubermaticapiv1.ClusterRoleBindingsCleanupFinalizer)
 	}
 
-	if !kuberneteshelper.HasFinalizer(cluster, kubermaticapiv1.ClusterRoleBindingsCleanupFinalizer) {
-		err := r.updateCluster(ctx, cluster, func(c *kubermaticv1.Cluster) {
-			kuberneteshelper.AddFinalizer(c, kubermaticapiv1.ClusterRoleBindingsCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
-		}
+	if len(finalizers) > 0 {
+		return r.AddFinalizers(ctx, cluster, finalizers...)
 	}
 
 	return &reconcile.Result{}, nil

--- a/pkg/provider/cloud/openstack/helper.go
+++ b/pkg/provider/cloud/openstack/helper.go
@@ -473,8 +473,8 @@ func isNotFoundErr(err error) bool {
 	return false
 }
 
-func getRouterIDForSubnet(netClient *gophercloud.ServiceClient, subnetID, networkID string) (string, error) {
-	ports, err := getAllNetworkPorts(netClient, networkID)
+func getRouterIDForSubnet(netClient *gophercloud.ServiceClient, subnetID string) (string, error) {
+	ports, err := getAllNetworkPorts(netClient, subnetID)
 	if err != nil {
 		return "", fmt.Errorf("failed to list ports for subnet: %v", err)
 	}
@@ -482,20 +482,16 @@ func getRouterIDForSubnet(netClient *gophercloud.ServiceClient, subnetID, networ
 	for _, port := range ports {
 		if port.DeviceOwner == "network:router_interface" || port.DeviceOwner == "network:router_interface_distributed" {
 			// Check IP for the interface & check if the IP belongs to the subnet
-			for _, ip := range port.FixedIPs {
-				if ip.SubnetID == subnetID {
-					return port.DeviceID, nil
-				}
-			}
+			return port.DeviceID, nil
 		}
 	}
 
 	return "", nil
 }
 
-func getAllNetworkPorts(netClient *gophercloud.ServiceClient, networkID string) ([]osports.Port, error) {
+func getAllNetworkPorts(netClient *gophercloud.ServiceClient, subnetID string) ([]osports.Port, error) {
 	allPages, err := osports.List(netClient, osports.ListOpts{
-		NetworkID: networkID,
+		FixedIPs: []osports.FixedIPOpts{{SubnetID: subnetID}},
 	}).AllPages()
 	if err != nil {
 		return nil, err

--- a/pkg/provider/cloud/openstack/internal/testing/fixtures.go
+++ b/pkg/provider/cloud/openstack/internal/testing/fixtures.go
@@ -1,0 +1,57 @@
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/external"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+)
+
+// IDs used when a new resource is created.
+// Note that using always the same ID is not correct, but is sufficient to
+// cover the current usage. For more advanced usages a pseudo-random UID
+// generator should be used instead.
+const (
+	NetworkID       = "8bb661f5-76b9-45f1-9ef9-eeffcd025fe4"
+	SubnetID        = "bec43a98-2d0a-4b1d-9df0-0f21e8f89d8a"
+	RouterID        = "b8a25073-b35f-4b64-b205-6d21c8221d98"
+	PortID          = "6f5c4730-aa3f-4677-bca2-3b1ead8545d9"
+	InterfaceInfoID = "6f5c4730-aa3f-4677-bca2-3b1ead8545d9"
+	SecGroupID      = "babf744f-0098-4f61-8e15-f91c30b3ed05"
+	SecGroupRuleID  = "da8d0577-eeb8-4351-b543-4bb0687f81d3"
+)
+
+// Relative paths to resource endpoints
+const (
+	SecurityGroupsEndpoint = "/security-groups"
+	NetworksEndpoint       = "/networks"
+	SubnetsEndpoint        = "/subnets"
+	RoutersEndpoint        = "/routers"
+)
+
+func AddRouterInterfaceEndpoint(routerID string) string {
+	return "/routers/" + routerID + "/add_router_interface"
+}
+
+var ExternalNetwork = Network{
+	Network:            networks.Network{Name: "external-network", ID: "d32019d3-bc6e-4319-9c1d-6722fc136a22"},
+	NetworkExternalExt: external.NetworkExternalExt{External: true},
+}
+
+var InternalNetwork = Network{
+	Network: networks.Network{Name: "kubernetes-cluster-xyz", ID: NetworkID},
+}

--- a/pkg/provider/cloud/openstack/internal/testing/simulator.go
+++ b/pkg/provider/cloud/openstack/internal/testing/simulator.go
@@ -1,0 +1,559 @@
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/external"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/layer3/routers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/groups"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/security/rules"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+	ostesthelper "github.com/gophercloud/gophercloud/testhelper"
+	osfakeclient "github.com/gophercloud/gophercloud/testhelper/client"
+)
+
+type Resource interface {
+	GetType() string
+	GetPath() string
+	GetID() string
+	GetName() string
+	FromCreateRequest([]byte) (Resource, error)
+	CreateResponse() ([]byte, error)
+	SubResources() []ResourceBuilder
+}
+
+type ResourceBuilder func() Resource
+
+type Network struct {
+	external.NetworkExternalExt
+	networks.Network
+}
+
+func (n Network) GetID() string {
+	return n.ID
+}
+
+func (n Network) GetName() string {
+	return n.Name
+}
+
+func (n Network) GetType() string {
+	return "networks"
+}
+
+func (n Network) GetPath() string {
+	return "/networks"
+}
+
+func (n *Network) FromCreateRequest(c []byte) (Resource, error) {
+	createOpts := struct {
+		networks.CreateOpts `json:"network"`
+	}{}
+	err := json.Unmarshal(c, &createOpts)
+	if err != nil {
+		return nil, err
+	}
+	n.ID = NetworkID
+	n.Name = createOpts.Name
+	n.Description = createOpts.Description
+	n.ProjectID = createOpts.ProjectID
+	n.UpdatedAt = time.Now()
+	n.CreatedAt = time.Now()
+	return n, nil
+}
+
+func (n *Network) CreateResponse() ([]byte, error) {
+	res := struct {
+		Network *Network `json:"network"`
+	}{
+		Network: n,
+	}
+	return json.Marshal(res)
+}
+
+func (n *Network) SubResources() []ResourceBuilder {
+	return nil
+}
+
+type SecGroup groups.SecGroup
+
+func (s SecGroup) GetID() string {
+	return s.ID
+}
+
+func (s SecGroup) GetName() string {
+	return s.Name
+}
+
+func (s SecGroup) GetType() string {
+	return "security_groups"
+}
+
+func (s SecGroup) GetPath() string {
+	return "/security-groups"
+}
+
+func (s *SecGroup) FromCreateRequest(c []byte) (Resource, error) {
+	createOpts := struct {
+		groups.CreateOpts `json:"security_group"`
+	}{}
+	err := json.Unmarshal(c, &createOpts)
+	if err != nil {
+		return nil, err
+	}
+	s.ID = SecGroupID
+	s.Name = createOpts.Name
+	s.Description = createOpts.Description
+	s.ProjectID = createOpts.ProjectID
+	s.UpdatedAt = time.Now()
+	s.CreatedAt = time.Now()
+	return s, nil
+}
+
+func (s *SecGroup) CreateResponse() ([]byte, error) {
+	res := struct {
+		SecGroup *SecGroup `json:"security_group"`
+	}{
+		SecGroup: s,
+	}
+	return json.Marshal(res)
+}
+
+func (s *SecGroup) SubResources() []ResourceBuilder {
+	return nil
+}
+
+type SecGroupRule rules.SecGroupRule
+
+func (s SecGroupRule) GetID() string {
+	return s.ID
+}
+
+func (s SecGroupRule) GetName() string {
+	return ""
+}
+
+func (s SecGroupRule) GetType() string {
+	return "security_group_rules"
+}
+
+func (s SecGroupRule) GetPath() string {
+	return "/security-group-rules"
+}
+
+func (s *SecGroupRule) FromCreateRequest(c []byte) (Resource, error) {
+	createOpts := struct {
+		rules.CreateOpts `json:"security_group_rule"`
+	}{}
+	err := json.Unmarshal(c, &createOpts)
+	if err != nil {
+		return nil, err
+	}
+	s.ID = SecGroupRuleID
+	s.Description = createOpts.Description
+	s.ProjectID = createOpts.ProjectID
+	s.Direction = string(createOpts.Direction)
+	s.EtherType = string(createOpts.EtherType)
+	s.PortRangeMax = createOpts.PortRangeMax
+	s.PortRangeMin = createOpts.PortRangeMin
+	return s, nil
+}
+
+func (s *SecGroupRule) CreateResponse() ([]byte, error) {
+	res := struct {
+		SecGroupRule *SecGroupRule `json:"security_group_rule"`
+	}{
+		SecGroupRule: s,
+	}
+	return json.Marshal(res)
+}
+
+func (s *SecGroupRule) SubResources() []ResourceBuilder {
+	return nil
+}
+
+type Subnet subnets.Subnet
+
+func (s Subnet) GetID() string {
+	return s.ID
+}
+
+func (s Subnet) GetName() string {
+	return s.Name
+}
+
+func (s Subnet) GetType() string {
+	return "subnets"
+}
+
+func (s Subnet) GetPath() string {
+	return "/subnets"
+}
+
+func (s *Subnet) FromCreateRequest(c []byte) (Resource, error) {
+	createOpts := struct {
+		subnets.CreateOpts `json:"subnet"`
+	}{}
+	err := json.Unmarshal(c, &createOpts)
+	if err != nil {
+		return nil, err
+	}
+	s.ID = SubnetID
+	s.Description = createOpts.Description
+	s.ProjectID = createOpts.ProjectID
+	s.AllocationPools = createOpts.AllocationPools
+	s.CIDR = createOpts.CIDR
+	s.DNSNameservers = createOpts.DNSNameservers
+	if createOpts.GatewayIP != nil {
+		s.GatewayIP = *createOpts.GatewayIP
+	}
+	if createOpts.EnableDHCP != nil {
+		s.EnableDHCP = *createOpts.EnableDHCP
+	}
+	return s, nil
+}
+
+func (s *Subnet) CreateResponse() ([]byte, error) {
+	res := struct {
+		Subnet *Subnet `json:"subnet"`
+	}{
+		Subnet: s,
+	}
+	return json.Marshal(res)
+}
+
+func (s *Subnet) SubResources() []ResourceBuilder {
+	return nil
+}
+
+type Port ports.Port
+
+func (p Port) GetID() string {
+	return p.ID
+}
+
+func (p Port) GetName() string {
+	return p.Name
+}
+
+func (p Port) GetType() string {
+	return "ports"
+}
+
+func (p Port) GetPath() string {
+	return "/ports"
+}
+
+func (p *Port) FromCreateRequest(c []byte) (Resource, error) {
+	createOpts := struct {
+		ports.CreateOpts `json:"port"`
+	}{}
+	err := json.Unmarshal(c, &createOpts)
+	if err != nil {
+		return nil, err
+	}
+	p.ID = PortID
+	p.Description = createOpts.Description
+	p.ProjectID = createOpts.ProjectID
+	p.DeviceID = createOpts.DeviceID
+	if createOpts.AdminStateUp != nil {
+		p.AdminStateUp = *createOpts.AdminStateUp
+	}
+	return p, nil
+}
+
+func (p *Port) CreateResponse() ([]byte, error) {
+	res := struct {
+		*Port `json:"subnet"`
+	}{
+		Port: p,
+	}
+	return json.Marshal(res)
+}
+
+func (p *Port) SubResources() []ResourceBuilder {
+	return nil
+}
+
+type Router routers.Router
+
+func (r Router) GetID() string {
+	return r.ID
+}
+
+func (r Router) GetName() string {
+	return r.Name
+}
+
+func (r Router) GetType() string {
+	return "routers"
+}
+
+func (r Router) GetPath() string {
+	return "/routers"
+}
+
+func (r *Router) FromCreateRequest(c []byte) (Resource, error) {
+	createOpts := struct {
+		routers.CreateOpts `json:"router"`
+	}{}
+	err := json.Unmarshal(c, &createOpts)
+	if err != nil {
+		return nil, err
+	}
+	r.ID = RouterID
+	r.Description = createOpts.Description
+	r.ProjectID = createOpts.ProjectID
+	if createOpts.Distributed != nil {
+		r.Distributed = *createOpts.Distributed
+	}
+	if createOpts.GatewayInfo != nil {
+		r.GatewayInfo = *createOpts.GatewayInfo
+	}
+	if createOpts.AdminStateUp != nil {
+		r.AdminStateUp = *createOpts.AdminStateUp
+	}
+	return r, nil
+}
+
+func (r *Router) CreateResponse() ([]byte, error) {
+	res := struct {
+		Router *Router `json:"router"`
+	}{
+		Router: r,
+	}
+	return json.Marshal(res)
+}
+
+func (r *Router) SubResources() []ResourceBuilder {
+	return []ResourceBuilder{
+		func() Resource {
+			return &InterfaceInfo{routerID: r.ID}
+		},
+	}
+}
+
+type InterfaceInfo struct {
+	routers.InterfaceInfo
+	routerID string
+}
+
+func (i InterfaceInfo) GetID() string {
+	return i.ID
+}
+
+func (i InterfaceInfo) GetName() string {
+	return ""
+}
+
+func (i InterfaceInfo) GetType() string {
+	return "interfacesInfo"
+}
+
+func (i InterfaceInfo) GetPath() string {
+	return fmt.Sprintf("/routers/%s/add_router_interface", i.routerID)
+}
+
+func (i *InterfaceInfo) FromCreateRequest(c []byte) (Resource, error) {
+	addOpts := routers.AddInterfaceOpts{}
+	err := json.Unmarshal(c, &addOpts)
+	if err != nil {
+		return nil, err
+	}
+	i.ID = InterfaceInfoID
+	i.PortID = addOpts.PortID
+	i.SubnetID = addOpts.SubnetID
+	return i, nil
+}
+
+func (i *InterfaceInfo) CreateResponse() ([]byte, error) {
+	return json.Marshal(i)
+}
+
+func (i *InterfaceInfo) SubResources() []ResourceBuilder {
+	return nil
+}
+
+type Request struct {
+	Method string
+	Path   string
+}
+
+// Simulator provides an OpenStack API server to be used for testing purposes.
+type Simulator struct {
+	*testing.T
+	sync.RWMutex
+	resources map[string][]Resource
+	builders  map[string]ResourceBuilder
+	reqCount  map[Request]int
+}
+
+func NewSimulator(t *testing.T) *Simulator {
+	// Start listening for incoming requests
+	ostesthelper.SetupHTTP()
+	// Create the simulator and register resource handlers
+	return (&Simulator{
+		T:         t,
+		builders:  map[string]ResourceBuilder{},
+		resources: map[string][]Resource{},
+		reqCount:  map[Request]int{},
+	}).
+		Register(func() Resource { return &Network{} }).
+		Register(func() Resource { return &Subnet{} }).
+		Register(func() Resource { return &SecGroup{} }).
+		Register(func() Resource { return &SecGroupRule{} }).
+		Register(func() Resource { return &Router{} }).
+		Register(func() Resource { return &Port{} })
+}
+
+func (s *Simulator) TearDown() {
+	ostesthelper.TeardownHTTP()
+}
+
+func (s *Simulator) Register(builder ResourceBuilder) *Simulator {
+	s.Lock()
+	defer s.Unlock()
+	path := builder().GetPath()
+	s.builders[path] = builder
+	ostesthelper.Mux.HandleFunc(path, s.Handle)
+	return s
+}
+
+func (s *Simulator) Add(res ...Resource) *Simulator {
+	for _, r := range res {
+		if b, ok := s.builders[r.GetPath()]; !ok || reflect.TypeOf(b()) != reflect.TypeOf(r) {
+			panic(fmt.Sprintf("cannot insert resource of type %s", r.GetType()))
+		}
+		s.resources[r.GetPath()] = append(s.resources[r.GetPath()], r)
+		for _, sub := range r.SubResources() {
+			s.Register(sub)
+		}
+	}
+	return s
+}
+
+func (s *Simulator) GetRequestCounters() map[Request]int {
+	s.RLock()
+	defer s.RUnlock()
+	currCounters := make(map[Request]int, len(s.reqCount))
+	for k, v := range s.reqCount {
+		currCounters[k] = v
+	}
+	return currCounters
+}
+
+func (s *Simulator) GetClient() *gophercloud.ServiceClient {
+	return osfakeclient.ServiceClient()
+}
+
+func (s *Simulator) Handle(w http.ResponseWriter, r *http.Request) {
+	// Increase req counter
+	s.Lock()
+	s.reqCount[Request{Path: r.URL.Path, Method: r.Method}]++
+	s.Unlock()
+	// Handle request
+	ostesthelper.TestHeader(s.T, r, "X-Auth-Token", osfakeclient.TokenID)
+	switch r.Method {
+	case http.MethodGet:
+		out, err := func() ([]byte, error) {
+			s.RLock()
+			defer s.RUnlock()
+			typeP := s.builders[r.URL.Path]().GetType()
+			name := r.URL.Query().Get("name")
+			// Filter by name.
+			var filtered []Resource
+			for _, r := range s.resources[r.URL.Path] {
+				if name != "" && r.GetName() != name {
+					continue
+				}
+				filtered = append(filtered, r)
+			}
+			return json.Marshal(map[string][]Resource{typeP: filtered})
+		}()
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, err.Error())
+		}
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, string(out))
+	case http.MethodPost:
+		reqBody, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, err.Error())
+		}
+		newRes, err := s.builders[r.URL.Path]().FromCreateRequest(reqBody)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, err.Error())
+		}
+		res, err := newRes.CreateResponse()
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, err.Error())
+		}
+		for _, dh := range newRes.SubResources() {
+			s.Register(dh)
+		}
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusCreated)
+		fmt.Fprintln(w, string(res))
+		s.Lock()
+		s.resources[r.URL.Path] = append(s.resources[r.URL.Path], newRes)
+		s.Unlock()
+	case http.MethodPut:
+		reqBody, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, err.Error())
+		}
+		newRes, err := s.builders[r.URL.Path]().FromCreateRequest(reqBody)
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, err.Error())
+		}
+		res, err := newRes.CreateResponse()
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintln(w, err.Error())
+		}
+		for _, dh := range newRes.SubResources() {
+			s.Register(dh)
+		}
+		w.Header().Add("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, string(res))
+		s.Lock()
+		s.resources[r.URL.Path] = append(s.resources[r.URL.Path], newRes)
+		s.Unlock()
+	default:
+		w.WriteHeader(http.StatusNotImplemented)
+	}
+}

--- a/pkg/provider/cloud/openstack/provider.go
+++ b/pkg/provider/cloud/openstack/provider.go
@@ -61,12 +61,15 @@ const (
 	RouterSubnetLinkCleanupFinalizer = "kubermatic.io/cleanup-openstack-router-subnet-link-v2"
 )
 
+type getClientFunc func(cluster kubermaticv1.CloudSpec, dc *kubermaticv1.DatacenterSpecOpenstack, secretKeySelector provider.SecretKeySelectorValueFunc, caBundle *x509.CertPool) (*gophercloud.ServiceClient, error)
+
 // Provider is a struct that implements CloudProvider interface
 type Provider struct {
 	dc                *kubermaticv1.DatacenterSpecOpenstack
 	secretKeySelector provider.SecretKeySelectorValueFunc
 	caBundle          *x509.CertPool
 	nodePortsRange    string
+	getClientFunc     getClientFunc
 }
 
 // NewCloudProvider creates a new openstack provider.
@@ -84,6 +87,7 @@ func NewCloudProvider(
 		secretKeySelector: secretKeyGetter,
 		caBundle:          caBundle,
 		nodePortsRange:    nodePortsRange,
+		getClientFunc:     getNetClientForCluster,
 	}, nil
 }
 
@@ -94,14 +98,9 @@ func (os *Provider) DefaultCloudSpec(spec *kubermaticv1.CloudSpec) error {
 
 // ValidateCloudSpec validates the given CloudSpec
 func (os *Provider) ValidateCloudSpec(spec kubermaticv1.CloudSpec) error {
-	creds, err := GetCredentialsForCluster(spec, os.secretKeySelector)
+	netClient, err := os.getClientFunc(spec, os.dc, os.secretKeySelector, os.caBundle)
 	if err != nil {
 		return err
-	}
-
-	netClient, err := getNetClient(creds.Username, creds.Password, creds.Domain, creds.Tenant, creds.TenantID, os.dc.AuthURL, os.dc.Region, os.caBundle)
-	if err != nil {
-		return fmt.Errorf("failed to create a authenticated openstack client: %v", err)
 	}
 
 	if spec.Openstack.SecurityGroups != "" {
@@ -168,14 +167,52 @@ func validateExistingSubnetOverlap(networkID string, netClient *gophercloud.Serv
 // InitializeCloudProvider initializes a cluster, in particular
 // creates security group and network configuration
 func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
-	creds, err := GetCredentialsForCluster(cluster.Spec.Cloud, os.secretKeySelector)
+
+	netClient, err := os.getClientFunc(cluster.Spec.Cloud, os.dc, os.secretKeySelector, os.caBundle)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get credentials: %v", err)
+		return nil, err
 	}
 
-	netClient, err := getNetClient(creds.Username, creds.Password, creds.Domain, creds.Tenant, creds.TenantID, os.dc.AuthURL, os.dc.Region, os.caBundle)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create a authenticated openstack client: %v", err)
+	var routerID string
+	var finalizers []string
+	// if security group has to be created add the corresponding finalizer.
+	if cluster.Spec.Cloud.Openstack.SecurityGroups == "" {
+		finalizers = append(finalizers, SecurityGroupCleanupFinalizer)
+	}
+	// If network has to be created add associated finalizer.
+	if cluster.Spec.Cloud.Openstack.Network == "" {
+		finalizers = append(finalizers, NetworkCleanupFinalizer)
+	}
+	// If subnet has to be created, router and router port should be
+	// created too thus we add the finalizers.
+	if cluster.Spec.Cloud.Openstack.SubnetID == "" {
+		finalizers = append(finalizers, SubnetCleanupFinalizer, RouterCleanupFinalizer, RouterSubnetLinkCleanupFinalizer)
+	} else if cluster.Spec.Cloud.Openstack.RouterID == "" {
+		var err error
+		routerID, err = getRouterIDForSubnet(netClient, cluster.Spec.Cloud.Openstack.SubnetID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to verify that the subnet '%s' has a router attached: %v", cluster.Spec.Cloud.Openstack.SubnetID, err)
+		}
+		// Subnet exists but we need to create the router
+		if routerID == "" {
+			finalizers = append(finalizers, RouterCleanupFinalizer, RouterSubnetLinkCleanupFinalizer)
+		}
+	}
+
+	// We start by adding the finalizers, note that this is safe because the
+	// clean-up is idempotent, if the cluster is deleted when resources associated
+	// to the finalizer are not created yet, it does not fail.
+	// The reason behind is that we have several controllers adding finalizers
+	// to the Cluster resource at the moment, and to avoid race conditions we
+	// need to use optimistic locking and return immediately in case of
+	// conflicts to retry later.
+	if len(finalizers) > 0 {
+		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
+			cluster.Finalizers = append(cluster.Finalizers, finalizers...)
+		}, provider.UpdaterOptionOptimisticLock)
+		if err != nil {
+			return nil, fmt.Errorf("failed to add finalizers: %w", err)
+		}
 	}
 
 	if cluster.Spec.Cloud.Openstack.FloatingIPPool == "" {
@@ -211,7 +248,6 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 		}
 		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
 			cluster.Spec.Cloud.Openstack.SecurityGroups = secGroupName
-			kubernetes.AddFinalizer(cluster, SecurityGroupCleanupFinalizer)
 		})
 		if err != nil {
 			return nil, fmt.Errorf("failed to add security group cleanup finalizer: %v", err)
@@ -225,7 +261,6 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 		}
 		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
 			cluster.Spec.Cloud.Openstack.Network = network.Name
-			kubernetes.AddFinalizer(cluster, NetworkCleanupFinalizer)
 		})
 		if err != nil {
 			return nil, fmt.Errorf("failed to add network cleanup finalizer: %v", err)
@@ -245,7 +280,6 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 
 		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
 			cluster.Spec.Cloud.Openstack.SubnetID = subnet.ID
-			kubernetes.AddFinalizer(cluster, SubnetCleanupFinalizer)
 		})
 		if err != nil {
 			return nil, fmt.Errorf("failed to add subnet cleanup finalizer: %v", err)
@@ -253,12 +287,6 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 	}
 
 	if cluster.Spec.Cloud.Openstack.RouterID == "" {
-		// Check if the subnet has already a router
-		routerID, err := getRouterIDForSubnet(netClient, cluster.Spec.Cloud.Openstack.SubnetID, network.ID)
-		if err != nil {
-			return nil, fmt.Errorf("failed to verify that the subnet '%s' has a router attached: %v", cluster.Spec.Cloud.Openstack.SubnetID, err)
-		}
-
 		if routerID == "" {
 			// No Router exists -> Create a router
 			router, err := createKubermaticRouter(netClient, cluster.Name, cluster.Spec.Cloud.Openstack.FloatingIPPool)
@@ -267,7 +295,6 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 			}
 			cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
 				cluster.Spec.Cloud.Openstack.RouterID = router.ID
-				kubernetes.AddFinalizer(cluster, RouterCleanupFinalizer)
 			})
 			if err != nil {
 				return nil, fmt.Errorf("failed to add router cleanup finalizer: %v", err)
@@ -283,18 +310,14 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 		}
 	}
 
-	// If we created the router, but have not created the router-subnet-link finalizer, we need to attach the subnet to the router
-	// Otherwise the vm's won't have connectivity
-	if kubernetes.HasFinalizer(cluster, RouterCleanupFinalizer) && !kubernetes.HasFinalizer(cluster, RouterSubnetLinkCleanupFinalizer) {
+	// We only attach the router to the subnet if CloudProviderInfrastructure
+	// health status is not up, meaning that there was no successful
+	// reconciliations so far. This is to avoid hitting OpenStack API at each
+	// iteration.
+	// TODO(irozzo): this is terrible, find a better way.
+	if cluster.Status.ExtendedHealth.CloudProviderInfrastructure != kubermaticv1.HealthStatusUp {
 		if _, err = attachSubnetToRouter(netClient, cluster.Spec.Cloud.Openstack.SubnetID, cluster.Spec.Cloud.Openstack.RouterID); err != nil {
 			return nil, fmt.Errorf("failed to attach subnet to router: %v", err)
-		}
-
-		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			kubernetes.AddFinalizer(cluster, RouterSubnetLinkCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, fmt.Errorf("failed to add router subnet cleanup finalizer: %v", err)
 		}
 	}
 
@@ -304,14 +327,9 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 // CleanUpCloudProvider does the clean-up in particular:
 // removes security group and network configuration
 func (os *Provider) CleanUpCloudProvider(cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
-	creds, err := GetCredentialsForCluster(cluster.Spec.Cloud, os.secretKeySelector)
+	netClient, err := os.getClientFunc(cluster.Spec.Cloud, os.dc, os.secretKeySelector, os.caBundle)
 	if err != nil {
 		return nil, err
-	}
-
-	netClient, err := getNetClient(creds.Username, creds.Password, creds.Domain, creds.Tenant, creds.TenantID, os.dc.AuthURL, os.dc.Region, os.caBundle)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create a authenticated openstack client: %v", err)
 	}
 
 	if kubernetes.HasFinalizer(cluster, SecurityGroupCleanupFinalizer) {
@@ -322,13 +340,6 @@ func (os *Provider) CleanUpCloudProvider(cluster *kubermaticv1.Cluster, update p
 				}
 			}
 		}
-
-		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			kubernetes.RemoveFinalizer(cluster, SecurityGroupCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
-		}
 	}
 
 	if kubernetes.HasFinalizer(cluster, RouterSubnetLinkCleanupFinalizer) || kubernetes.HasFinalizer(cluster, OldNetworkCleanupFinalizer) {
@@ -336,12 +347,6 @@ func (os *Provider) CleanUpCloudProvider(cluster *kubermaticv1.Cluster, update p
 			if !isNotFoundErr(err) {
 				return nil, fmt.Errorf("failed to detach subnet from router: %v", err)
 			}
-		}
-		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			kubernetes.RemoveFinalizer(cluster, RouterSubnetLinkCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
 		}
 	}
 
@@ -351,12 +356,6 @@ func (os *Provider) CleanUpCloudProvider(cluster *kubermaticv1.Cluster, update p
 				return nil, fmt.Errorf("failed to delete subnet '%s': %v", cluster.Spec.Cloud.Openstack.SubnetID, err)
 			}
 		}
-		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			kubernetes.RemoveFinalizer(cluster, SubnetCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
-		}
 	}
 
 	if kubernetes.HasFinalizer(cluster, NetworkCleanupFinalizer) || kubernetes.HasFinalizer(cluster, OldNetworkCleanupFinalizer) {
@@ -364,13 +363,6 @@ func (os *Provider) CleanUpCloudProvider(cluster *kubermaticv1.Cluster, update p
 			if !isNotFoundErr(err) {
 				return nil, fmt.Errorf("failed to delete network '%s': %v", cluster.Spec.Cloud.Openstack.Network, err)
 			}
-		}
-
-		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			kubernetes.RemoveFinalizer(cluster, NetworkCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
 		}
 	}
 
@@ -380,22 +372,23 @@ func (os *Provider) CleanUpCloudProvider(cluster *kubermaticv1.Cluster, update p
 				return nil, fmt.Errorf("failed to delete router '%s': %v", cluster.Spec.Cloud.Openstack.RouterID, err)
 			}
 		}
-
-		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			kubernetes.RemoveFinalizer(cluster, RouterCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
-		}
 	}
 
-	if kubernetes.HasFinalizer(cluster, OldNetworkCleanupFinalizer) {
-		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			kubernetes.RemoveFinalizer(cluster, OldNetworkCleanupFinalizer)
-		})
-		if err != nil {
-			return nil, err
-		}
+	// Relying on the idempotence of the clean-up steps we remove all finalizers in
+	// one shot only when the clean-up is completed.
+	cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
+		kubernetes.RemoveFinalizer(
+			cluster,
+			SecurityGroupCleanupFinalizer,
+			RouterSubnetLinkCleanupFinalizer,
+			SubnetCleanupFinalizer,
+			NetworkCleanupFinalizer,
+			RouterCleanupFinalizer,
+			OldNetworkCleanupFinalizer,
+		)
+	}, provider.UpdaterOptionOptimisticLock)
+	if err != nil {
+		return nil, err
 	}
 
 	return cluster, nil
@@ -639,6 +632,19 @@ func addICMPRulesToSecurityGroupIfNecesary(cluster *kubermaticv1.Cluster, secGro
 // ValidateCloudSpecUpdate verifies whether an update of cloud spec is valid and permitted
 func (os *Provider) ValidateCloudSpecUpdate(oldSpec kubermaticv1.CloudSpec, newSpec kubermaticv1.CloudSpec) error {
 	return nil
+}
+
+func getNetClientForCluster(cluster kubermaticv1.CloudSpec, dc *kubermaticv1.DatacenterSpecOpenstack, secretKeySelector provider.SecretKeySelectorValueFunc, caBundle *x509.CertPool) (*gophercloud.ServiceClient, error) {
+	creds, err := GetCredentialsForCluster(cluster, secretKeySelector)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get credentials: %v", err)
+	}
+
+	netClient, err := getNetClient(creds.Username, creds.Password, creds.Domain, creds.Tenant, creds.TenantID, dc.AuthURL, dc.Region, caBundle)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a authenticated openstack client: %v", err)
+	}
+	return netClient, nil
 }
 
 // GetCredentialsForCluster returns the credentials for the passed in cloud spec or an error

--- a/pkg/provider/cloud/openstack/provider.go
+++ b/pkg/provider/cloud/openstack/provider.go
@@ -208,7 +208,7 @@ func (os *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, updat
 	// conflicts to retry later.
 	if len(finalizers) > 0 {
 		cluster, err = update(cluster.Name, func(cluster *kubermaticv1.Cluster) {
-			cluster.Finalizers = append(cluster.Finalizers, finalizers...)
+			kubernetes.AddFinalizer(cluster, finalizers...)
 		}, provider.UpdaterOptionOptimisticLock)
 		if err != nil {
 			return nil, fmt.Errorf("failed to add finalizers: %w", err)

--- a/pkg/provider/cloud/openstack/provider_test.go
+++ b/pkg/provider/cloud/openstack/provider_test.go
@@ -17,9 +17,18 @@ limitations under the License.
 package openstack
 
 import (
+	"crypto/x509"
+	"net/http"
 	"testing"
 
+	"github.com/go-test/deep"
 	"github.com/gophercloud/gophercloud"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/provider"
+	ostesting "k8c.io/kubermatic/v2/pkg/provider/cloud/openstack/internal/testing"
 )
 
 func TestIgnoreRouterAlreadyHasPortInSubnetError(t *testing.T) {
@@ -56,4 +65,308 @@ func TestIgnoreRouterAlreadyHasPortInSubnetError(t *testing.T) {
 			}
 		})
 	}
+}
+
+func TestInitializeCloudProvider(t *testing.T) {
+	tests := []struct {
+		name         string
+		dc           *kubermaticv1.DatacenterSpecOpenstack
+		cluster      *kubermaticv1.Cluster
+		resources    []ostesting.Resource
+		wantErr      bool
+		wantCluster  kubermaticv1.Cluster
+		wantRequests map[ostesting.Request]int
+	}{
+		{
+			name: "Create all",
+			dc:   &kubermaticv1.DatacenterSpecOpenstack{},
+			cluster: &kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{},
+					},
+				},
+			},
+			resources: []ostesting.Resource{&ostesting.ExternalNetwork},
+			wantCluster: kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+					Finalizers: []string{
+						SecurityGroupCleanupFinalizer,
+						NetworkCleanupFinalizer,
+						SubnetCleanupFinalizer,
+						RouterCleanupFinalizer,
+						RouterSubnetLinkCleanupFinalizer,
+					},
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							SecurityGroups: "kubernetes-cluster-xyz",
+							FloatingIPPool: "external-network",
+							Network:        "kubernetes-cluster-xyz",
+							SubnetID:       ostesting.SubnetID,
+							RouterID:       ostesting.RouterID,
+						},
+					},
+				},
+			},
+			wantErr: false,
+			wantRequests: map[ostesting.Request]int{
+				{Method: http.MethodPost, Path: ostesting.SecurityGroupsEndpoint}:                        1,
+				{Method: http.MethodPost, Path: ostesting.NetworksEndpoint}:                              1,
+				{Method: http.MethodPost, Path: ostesting.SubnetsEndpoint}:                               1,
+				{Method: http.MethodPost, Path: ostesting.RoutersEndpoint}:                               1,
+				{Method: http.MethodPut, Path: ostesting.AddRouterInterfaceEndpoint(ostesting.RouterID)}: 1,
+			},
+		},
+		{
+			name: "Create nothing",
+			dc:   &kubermaticv1.DatacenterSpecOpenstack{},
+			cluster: &kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							SecurityGroups: "kubernetes-cluster-xyz",
+							FloatingIPPool: "external-network",
+							Network:        "kubernetes-cluster-xyz",
+							SubnetID:       ostesting.SubnetID,
+							RouterID:       ostesting.RouterID,
+						},
+					},
+				},
+				Status: kubermaticv1.ClusterStatus{
+					ExtendedHealth: kubermaticv1.ExtendedClusterHealth{
+						CloudProviderInfrastructure: kubermaticv1.HealthStatusUp,
+					},
+				},
+			},
+			resources: []ostesting.Resource{
+				&ostesting.ExternalNetwork,
+				&ostesting.InternalNetwork,
+				&ostesting.Router{
+					ID: ostesting.RouterID,
+				},
+			},
+			wantCluster: kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							SecurityGroups: "kubernetes-cluster-xyz",
+							FloatingIPPool: "external-network",
+							Network:        "kubernetes-cluster-xyz",
+							SubnetID:       ostesting.SubnetID,
+							RouterID:       ostesting.RouterID,
+						},
+					},
+				},
+				Status: kubermaticv1.ClusterStatus{
+					ExtendedHealth: kubermaticv1.ExtendedClusterHealth{
+						CloudProviderInfrastructure: kubermaticv1.HealthStatusUp,
+					},
+				},
+			},
+			wantRequests: map[ostesting.Request]int{
+				{Method: http.MethodPost, Path: ostesting.SecurityGroupsEndpoint}:                        0,
+				{Method: http.MethodPost, Path: ostesting.NetworksEndpoint}:                              0,
+				{Method: http.MethodPost, Path: ostesting.SubnetsEndpoint}:                               0,
+				{Method: http.MethodPost, Path: ostesting.RoutersEndpoint}:                               0,
+				{Method: http.MethodPut, Path: ostesting.AddRouterInterfaceEndpoint(ostesting.RouterID)}: 0,
+			},
+			wantErr: false,
+		},
+		{
+			name: "Network provided",
+			dc:   &kubermaticv1.DatacenterSpecOpenstack{},
+			cluster: &kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							Network: "kubernetes-cluster-xyz",
+						},
+					},
+				},
+			},
+			resources: []ostesting.Resource{
+				&ostesting.ExternalNetwork,
+				&ostesting.InternalNetwork,
+			},
+			wantCluster: kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+					Finalizers: []string{
+						SecurityGroupCleanupFinalizer,
+						SubnetCleanupFinalizer,
+						RouterCleanupFinalizer,
+						RouterSubnetLinkCleanupFinalizer,
+					},
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							SecurityGroups: "kubernetes-cluster-xyz",
+							FloatingIPPool: "external-network",
+							Network:        "kubernetes-cluster-xyz",
+							SubnetID:       ostesting.SubnetID,
+							RouterID:       ostesting.RouterID,
+						},
+					},
+				},
+			},
+			wantErr: false,
+			wantRequests: map[ostesting.Request]int{
+				{Method: http.MethodPost, Path: ostesting.SecurityGroupsEndpoint}:                        1,
+				{Method: http.MethodPost, Path: ostesting.NetworksEndpoint}:                              0,
+				{Method: http.MethodPost, Path: ostesting.SubnetsEndpoint}:                               1,
+				{Method: http.MethodPost, Path: ostesting.RoutersEndpoint}:                               1,
+				{Method: http.MethodPut, Path: ostesting.AddRouterInterfaceEndpoint(ostesting.RouterID)}: 1,
+			},
+		},
+		{
+			name: "Network and subnet provided",
+			dc:   &kubermaticv1.DatacenterSpecOpenstack{},
+			cluster: &kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							Network:  "kubernetes-cluster-xyz",
+							SubnetID: ostesting.SubnetID,
+						},
+					},
+				},
+			},
+			resources: []ostesting.Resource{
+				&ostesting.ExternalNetwork,
+				&ostesting.InternalNetwork,
+			},
+			wantCluster: kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+					Finalizers: []string{
+						SecurityGroupCleanupFinalizer,
+						RouterCleanupFinalizer,
+						RouterSubnetLinkCleanupFinalizer,
+					},
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							SecurityGroups: "kubernetes-cluster-xyz",
+							FloatingIPPool: "external-network",
+							Network:        "kubernetes-cluster-xyz",
+							SubnetID:       ostesting.SubnetID,
+							RouterID:       ostesting.RouterID,
+						},
+					},
+				},
+			},
+			wantErr: false,
+			wantRequests: map[ostesting.Request]int{
+				{Method: http.MethodPost, Path: ostesting.SecurityGroupsEndpoint}:                        1,
+				{Method: http.MethodPost, Path: ostesting.NetworksEndpoint}:                              0,
+				{Method: http.MethodPost, Path: ostesting.SubnetsEndpoint}:                               0,
+				{Method: http.MethodPost, Path: ostesting.RoutersEndpoint}:                               1,
+				{Method: http.MethodPut, Path: ostesting.AddRouterInterfaceEndpoint(ostesting.RouterID)}: 1,
+			},
+		},
+		{
+			name: "Specified network not found",
+			dc:   &kubermaticv1.DatacenterSpecOpenstack{},
+			cluster: &kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							Network:  "kubernetes-cluster-xyz",
+							SubnetID: ostesting.SubnetID,
+						},
+					},
+				},
+			},
+			resources: []ostesting.Resource{
+				&ostesting.ExternalNetwork,
+			},
+			wantCluster: kubermaticv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "cluster-xyz",
+					Finalizers: []string{
+						SecurityGroupCleanupFinalizer,
+						RouterCleanupFinalizer,
+						RouterSubnetLinkCleanupFinalizer,
+					},
+				},
+				Spec: kubermaticv1.ClusterSpec{
+					Cloud: kubermaticv1.CloudSpec{
+						Openstack: &kubermaticv1.OpenstackCloudSpec{
+							SecurityGroups: "kubernetes-cluster-xyz",
+							FloatingIPPool: "external-network",
+							Network:        "kubernetes-cluster-xyz",
+							SubnetID:       ostesting.SubnetID,
+							RouterID:       ostesting.RouterID,
+						},
+					},
+				},
+			},
+			wantErr: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := ostesting.NewSimulator(t).Add(tt.resources...)
+			defer s.TearDown()
+
+			os := &Provider{
+				dc: tt.dc,
+				getClientFunc: func(cluster kubermaticv1.CloudSpec, dc *kubermaticv1.DatacenterSpecOpenstack, secretKeySelector provider.SecretKeySelectorValueFunc, caBundle *x509.CertPool) (*gophercloud.ServiceClient, error) {
+					sc := s.GetClient()
+					return sc, nil
+				},
+			}
+			c, err := os.InitializeCloudProvider(tt.cluster, (&fakeClusterUpdater{c: tt.cluster}).update)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Provider.InitializeCloudProvider() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			// No need to proceed with further tests if an error is expected.
+			if tt.wantErr {
+				return
+			}
+			if diff := deep.Equal(tt.wantCluster, *c); len(diff) > 0 {
+				t.Errorf("Diff found between actual and wanted cluster: %v", diff)
+			}
+			rc := s.GetRequestCounters()
+			for req, e := range tt.wantRequests {
+				if a := rc[req]; a != e {
+					t.Errorf("Wanted %d requests %s, but got %d", e, req, a)
+				}
+			}
+		})
+	}
+}
+
+type fakeClusterUpdater struct {
+	c *kubermaticv1.Cluster
+}
+
+func (f *fakeClusterUpdater) update(_ string, updateFn func(c *kubermaticv1.Cluster), _ ...provider.UpdaterOption) (*kubermaticv1.Cluster, error) {
+	updateFn(f.c)
+	return f.c, nil
 }

--- a/pkg/provider/cloud/openstack/provider_test.go
+++ b/pkg/provider/cloud/openstack/provider_test.go
@@ -24,11 +24,12 @@ import (
 	"github.com/go-test/deep"
 	"github.com/gophercloud/gophercloud"
 
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-
 	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
 	"k8c.io/kubermatic/v2/pkg/provider"
 	ostesting "k8c.io/kubermatic/v2/pkg/provider/cloud/openstack/internal/testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
 )
 
 func TestIgnoreRouterAlreadyHasPortInSubnetError(t *testing.T) {
@@ -349,6 +350,15 @@ func TestInitializeCloudProvider(t *testing.T) {
 			if tt.wantErr {
 				return
 			}
+			// We cannot guarantee order if finalizers serialized using sets and deep.Equal fails if order is different,
+			// thus we test finalizers equality separately.
+			if w, g := sets.NewString(tt.wantCluster.Finalizers...), sets.NewString(c.Finalizers...); !w.Equal(g) {
+				t.Errorf("Want finalizers: %v, got: %v", w, g)
+			} else {
+				tt.wantCluster.Finalizers = nil
+				c.Finalizers = nil
+			}
+
 			if diff := deep.Equal(tt.wantCluster, *c); len(diff) > 0 {
 				t.Errorf("Diff found between actual and wanted cluster: %v", diff)
 			}

--- a/pkg/provider/types.go
+++ b/pkg/provider/types.go
@@ -76,8 +76,34 @@ type CloudProvider interface {
 	ValidateCloudSpecUpdate(oldSpec kubermaticv1.CloudSpec, newSpec kubermaticv1.CloudSpec) error
 }
 
+// UpdaterOption represent an option for the updater function.
+type UpdaterOption string
+
+const (
+	// UpdaterOptionOptimisticLock enables optimistic lock, to fail in case of
+	// potential conflict.
+	UpdaterOptionOptimisticLock UpdaterOption = "OptimisticLock"
+)
+
+// UpdaterOptions holds the options for the updater function.
+type UpdaterOptions struct {
+	OptimisticLock bool
+}
+
+func (c *UpdaterOptions) Apply(opts ...UpdaterOption) *UpdaterOptions {
+	for _, o := range opts {
+		switch o {
+		case UpdaterOptionOptimisticLock:
+			c.OptimisticLock = true
+		default:
+			panic(fmt.Sprintf("unrecognised cluster updater option: %s", o))
+		}
+	}
+	return c
+}
+
 // ClusterUpdater defines a function to persist an update to a cluster
-type ClusterUpdater func(string, func(*kubermaticv1.Cluster)) (*kubermaticv1.Cluster, error)
+type ClusterUpdater func(string, func(*kubermaticv1.Cluster), ...UpdaterOption) (*kubermaticv1.Cluster, error)
 
 // ClusterListOptions allows to set filters that will be applied to filter the result.
 type ClusterListOptions struct {

